### PR TITLE
port SetIPProtectionLevel and AllowNatTraversal

### LINKED ISSUE #12462: Add back Socket.SetIPProtectionLevel and UdpClient.AllowNatTraversal
```csharp
    public class Socket : IDisposable {
        public void SetIPProtectionLevel(IPProtectionLevel level);
    }
    public class UdpClient : IDisposable {
        public void AllowNatTraversal(bool allowed);
    }
```

(Was tracked by #11788)

### PR DESCRIPTION
**Pull request by geoffkizer**
Fixes #12462 

Add Socket.SetIPProtectionLevel.  Not supported on Unix (throws PlatformNotSupportedException)
Add UdpClient.AllowNatTraversal and TcpListener.AllowNatTraversal which use this.

### CODE REVIEW

**Comment by geoffkizer** [comment]
Test Outerloop Windows_NT Debug please
Test Outerloop OSX Debug please
Test Outerloop Ubuntu14.04 Debug please
Test Outerloop CentOS7.1 Debug please

**Comment by geoffkizer** [comment]
@stephentoub @ericeil @davidsh @CIPop

File: src/System.Net.Sockets/src/System/Net/Sockets/TCPListener.cs [hunk 1]
```diff
@@ -123,6 +123,23 @@ public bool ExclusiveAddressUse
...
+            }
+
+            if (allowed)
+            {
+                _serverSocket.SetIPProtectionLevel(IPProtectionLevel.Unrestricted);
+            }
+            else
+            {
+                _serverSocket.SetIPProtectionLevel(IPProtectionLevel.EdgeRestricted);
+            }
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/TCPListener.cs]: Nit: could be:
```C#
_serverSocket.SetIPProtectionLevel(allowed ? IPProtectionLevel.Unrestricted : IPProtectionLevel.EdgeRestricted);
```

File: src/System.Net.Sockets/src/System/Net/Sockets/UDPClient.cs [hunk 2]
```diff
@@ -194,6 +194,18 @@ public bool ExclusiveAddressUse
...
+        public void AllowNatTraversal(bool allowed)
+        {
+            if (allowed)
+            {
+                _clientSocket.SetIPProtectionLevel(IPProtectionLevel.Unrestricted);
+            }
+            else
+            {
+                _clientSocket.SetIPProtectionLevel(IPProtectionLevel.EdgeRestricted);
+            }
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/UDPClient.cs]: Ditto

File: src/System.Net.Sockets/tests/FunctionalTests/AgnosticListenerTest.cs [hunk 3]
```diff
@@ -112,6 +112,33 @@ public void StaticCreate_Success()
...
+        [Fact]
+        [PlatformSpecific(TestPlatforms.AnyUnix)]
+        public void AllowNatTraversal_AnyUnix()
+        {
+            var l = new TcpListener(IPAddress.Any, 0);
+            Assert.Throws<PlatformNotSupportedException>(() => l.AllowNatTraversal(true));
+
+            l = new TcpListener(IPAddress.Any, 0);
+            Assert.Throws<PlatformNotSupportedException>(() => l.AllowNatTraversal(false));
+        }
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/AgnosticListenerTest.cs]: Nit: if you make this a Theory with two InlineDatas, one for true and one for false, you can avoid the duplication, e.g.
```C#
[OuterLoop]
[PlatformSpecific(TestPlatforms.AnyUnix)]
[Theory]
[InlineData(true)]
[InlineData(false)]
public void AllowNatTraversal_AnyUnix(bool allowed)
{
    var listener = new TcpListener(IPAddress.Any, 0);
    Assert.Throws<PlatformNotSupportedException>(() => l.AllowNatTraversal(allowed));
}
```

File: src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs [hunk 4]
```diff
@@ -293,5 +293,64 @@ public void ReuseAddress_Windows(bool? exclusiveAddressUse, bool? firstSocketReu
...
+            }
+
+            using (var socket = new Socket(AddressFamily.InterNetworkV6, SocketType.Stream, ProtocolType.Tcp))
+            {
+                socket.SetIPProtectionLevel(level);
+
+                int result = (int)socket.GetSocketOption(SocketOptionLevel.IPv6, SocketOptionName.IPProtectionLevel);
+                Assert.Equal(result, (int)level);
+            }
+        }
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs]: Is there anything more we can/should test about the impact this option has on behavior?  If not, ok.

File: src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs [hunk 5]
```diff
@@ -293,5 +293,64 @@ public void ReuseAddress_Windows(bool? exclusiveAddressUse, bool? firstSocketReu
...
+        {
+            using (var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                socket.SetIPProtectionLevel(level);
+
+                int result = (int)socket.GetSocketOption(SocketOptionLevel.IP, SocketOptionName.IPProtectionLevel);
+                Assert.Equal(result, (int)level);
+            }
+
+            using (var socket = new Socket(AddressFamily.InterNetworkV6, SocketType.Stream, ProtocolType.Tcp))
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs]: As in my earlier comment about a theory, you could avoid the duplication by including the AddressFamily in the test args.

File: src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 6]
```diff
@@ -2276,6 +2276,27 @@ public byte[] GetSocketOption(SocketOptionLevel optionLevel, SocketOptionName op
             return optionValue;
         }
 
+        public void SetIPProtectionLevel(IPProtectionLevel level)
+        {
+            if (level == IPProtectionLevel.Unspecified)
+            {
+                throw new ArgumentException(SR.Format(SR.net_sockets_invalid_optionValue_all), "level");
```
Comment (justinvp) [inline src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs]: Nit: nameof(level)

File: src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs [hunk 7]
```diff
@@ -293,5 +293,64 @@ public void ReuseAddress_Windows(bool? exclusiveAddressUse, bool? firstSocketReu
...
+            }
+        }
+
+        [OuterLoop] // TODO: Issue #11345
+        [Fact]
+        public void SetIPProtectionLevel_ArgumentException()
+        {
+            using (var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                Assert.Throws<ArgumentException>(() => socket.SetIPProtectionLevel(IPProtectionLevel.Unspecified));
```
Comment (justinvp) [inline src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs]: Nit: here and below: `Assert.Throws<ArgumentException>("level", () => ...);`

**Comment by geoffkizer** [comment]
Test Outerloop Windows_NT Debug please
Test Outerloop OSX Debug please
Test Outerloop Ubuntu14.04 Debug please
Test Outerloop CentOS7.1 Debug please

File: src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs [hunk 8]
```diff
@@ -131,7 +131,7 @@ public void MulticastOption_CreateSocketSetGetOption_GroupAndInterfaceIndex_SetS
             int port;
 
             using (Socket receiveSocket = CreateBoundUdpSocket(out port),
-                          sendSocket    = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp))
+                          sendSocket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp))
```
Comment (CIPop) [inline src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs]: Nit: I think multiple `using` statements look better then using comma.

**Comment by geoffkizer** [comment]
Test Windows_NT Release please

**Comment by geoffkizer** [comment]
Test Outerloop Windows_NT Debug please

**Comment by geoffkizer** [comment]
test this please

**Comment by geoffkizer** [comment]
test Innerloop OSX Debug Build and Test please

**Comment by geoffkizer** [comment]
@dotnet-bot test Innerloop OSX Debug Build and Test

**Comment by stephentoub** [comment]
@dotnet-bot Test OuterLoop Windows_NT Debug please (https://github.com/dotnet/corefx/issues/12467)

**Comment by stephentoub** [comment]
@dotnet-bot Test OuterLoop Windows_NT Debug please (#12467)

**Comment by geoffkizer** [comment]
@dotnet-bot Test OuterLoop Windows_NT Debug

**Pull request merged**